.github/workflows/build-and-test.yml

```diff
@@ -0,0 +1,145 @@
+
+  test-suite:
+    strategy:
+      matrix:
+        java: ['17', '11', '8']
+    runs-on: ubuntu-latest
+    name: Run test suite with Java ${{ matrix.java }}
+    needs: build-and-check
+    steps:
+      - uses: actions/checkout@v4
+
+      - uses: actions/setup-java@v4
+        name: Set up Java
+        with:
+          java-version: ${{ matrix.java }}
+          distribution: 'temurin'
+
+      - name: Suid root tar
+        # Required for the bats libraries cache to restore properly
+        run: sudo chmod +s /bin/tar
+
+      - name: Setup Bats and bats libs
+        id: setup-bats
+        uses: bats-core/bats-action@3.0.0
+
+      - name: Remove suid root tar
+        run: sudo chmod -s /bin/tar
+
+      - uses: sbt/setup-sbt@v1
+
+      - name: Download artifact
+        uses: actions/download-artifact@v4
+        with:
+          name: Jars
+
+      - name: Clone test projects
+        shell: bash
+        env:
+          BATS_LIB_PATH: ${{ steps.setup-bats.outputs.lib-path }}
+          TERM: xterm
+        working-directory: ./agent
+        run: bin/test_projects
+
+      - uses: actions/setup-java@v4
+        name: Set up Gradle cache
+        with:
+          java-version: ${{ matrix.java }}
+          distribution: 'temurin'
+          cache: gradle
+
+      - uses: actions/setup-java@v4
+        name: Set up Maven cache
+        with:
+          java-version: ${{ matrix.java }}
+          distribution: 'temurin'
+          cache: maven
+
+      - uses: actions/setup-java@v4
+        name: Set up sbt cache
+        with:
+          java-version: ${{ matrix.java }}
+          distribution: 'temurin'
+          cache: sbt
+
+      - name: Build test projects
+        shell: bash
+        env:
+          BATS_LIB_PATH: ${{ steps.setup-bats.outputs.lib-path }}
+          TERM: xterm
+        working-directory: ./agent
+        run: bin/test_install
+
+      - name: Run tests
+        shell: bash
+        env:
+          BATS_LIB_PATH: ${{ steps.setup-bats.outputs.lib-path }}
+          TERM: xterm
+        working-directory: ./agent
+        run: bin/test_run
+
+  release:
+    name: Release
+    runs-on: ubuntu-latest
+    needs: test-suite
+    if: ${{ github.ref == 'refs/heads/master' }}
+    steps:
+      - uses: actions/checkout@v4
+      - uses: actions/setup-java@v4
+        with:
+          java-version: '17'
+          distribution: 'temurin'
+      - name: Setup Gradle
+        uses: gradle/actions/setup-gradle@v4
+      - uses: actions/download-artifact@v4
+        with:
+          name: Jars
+      - name: Install semantic-release
+        run: |
+          npm i -g \
+              semantic-release \
+              @semantic-release/exec \
+              @semantic-release/git \
+              @semantic-release/changelog \
+              @google/semantic-release-replace-plugin
+      - name: Run semantic-release
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```

Comment:
The automatic `GITHUB_TOKEN` only has read permissions, so I'm adding a new token to this repositories secret store. We can't name secrets anything starting with `GITHUB_` so I'm calling it `GH_TOKEN`.